### PR TITLE
Tuple Predicate Embedder

### PR DESCRIPTION
**Pull request by thesamuel**
Added a predicate embedder to help us evaluate new tuples.

We can use cosine similarity between the embeddings generated for new tuples and all other predicate embeddings to get an idea of how good each tuple is.

To train, run this from the project root:
```
allennlp train tuple_filtering/tuple_predicate_embedder.json -s /tmp/train --include-package tuple_filtering 
```

### CODE REVIEW

File: tuple_filtering/tuple_predicate_embedder.json [hunk 1]
```diff
@@ -0,0 +1,95 @@
+{
```
Comment (thesamuel) [inline tuple_filtering/tuple_predicate_embedder.json:1]: Here's the AllenNLP config I wrote

File: tuple_filtering/tuple_predicate_embedder.py [hunk 2]
```diff
@@ -0,0 +1,67 @@
+from typing import Dict, Optional
```
Comment (thesamuel) [inline tuple_filtering/tuple_predicate_embedder.py:1]: And here's the embedder model itself...

File: allennlp_runner.py [hunk 3]
```diff
@@ -0,0 +1,32 @@
+import json
```
Comment (thesamuel) [inline allennlp_runner.py:1]: This is a runner for the tuple embedder (can be used for pretty much any other AllenNLP task we have)

File: tuple_filtering/tuple_predicate_embedder.py [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+        if predicate_tokens:
+            gold_embedding = self.embed_predicate(predicate_tokens)
+
+            # Compute cosine loss between gold embedding and outputted embedding
+            cosine_loss_label = torch.tensor([1], dtype=out_embedding.dtype, device=out_embedding.device)
+            output_dict["loss"] = F.cosine_embedding_loss(out_embedding, gold_embedding, cosine_loss_label)
+
+        return output_dict
+
+    def embed_predicate(self, predicate_tokens: Dict[str, torch.LongTensor]):
```
Comment (thesamuel) [inline tuple_filtering/tuple_predicate_embedder.py:66]: After re-loading a saved model, we could embed every predicate from Aristo. Using those embeddings, we could go through every tuple in OpenBookQA and see if there's a close match for its predicate.

One thing that I overlooked is that we aren't using any predicate information during training (other than for computing the loss). Maybe we can find a way to incorporate the predicate from OpenBook into our model....
